modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/SchemaManager.java

```diff
@@ -616,6 +616,12 @@ public void dropIndex(final String schemaName, String idxName, boolean ifExists)
         throws IgniteCheckedException{
         String sql = H2Utils.indexDropSql(schemaName, idxName, ifExists);
 
+        GridH2Table tbl = dataTableForIndex(schemaName, idxName);
+
+        assert tbl != null;
+
+        tbl.setRemoveIndexOnDestroy(true);
```

Comment:
I am thinking about a race condition when an index is being dropped and a cache is stopping on deactivation concurrently. A following invalid scenario looks possible:
1. On deactivation `rmIndex` flag is set to false.
2. On index drop `rmIndex`is set to true. Index is dropped.
3. `rmIndex == true` leads that all indexes are destroyed on deactivation.

Reply:
Also a method name `setRemoveIndexOnDestroy` (and a javadoc) becomes misleading as it is going to be used for destroying an index on drop.

Reply:
Ivan, 
Cache stop and  schema operations (e.g. index drop) causes PME and should be serialized.
All schema operations are queued via GridQueryProcessor.schemaOps.
Cache stop operation fall into GridQueryProcessor.onCacheStop0 where all schemaOps have cancelled right before 
setRemoveIndexOnDestroy flag is changed (inside idx.unregisterCache() call).

Possibly, GridQueryProcessor should wait for active SchemaOperationManager.worker-s before idx.unregisterCache(), but this should be investigated.

Reply:
> Possibly, GridQueryProcessor should wait for active SchemaOperationManager.worker-s before idx.unregisterCache(), but this should be investigated.
Will be good to check it. AFAIK, cancellation uses `Thread.interrupt`, which cannot cancel running `setRemoveIndexOnDestroy` method alone.

Also, I think that it would be great to separate an intent of _removing_ indexes on per-operation basis rather than having a single boolean flag shared between different operations (cache stop, drop index).

Reply:
I wonder, why do we need to do fake "Drop table" on deactivation, though database "Shutdown" is called later?